server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java

```diff
@@ -121,23 +126,41 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
-    private boolean checkIfSameWeightsInMetadata(
-        WeightedRoutingMetadata newWeightedRoutingMetadata,
-        WeightedRoutingMetadata oldWeightedRoutingMetadata
-    ) {
-        return newWeightedRoutingMetadata.getWeightedRouting().equals(oldWeightedRoutingMetadata.getWeightedRouting());
+    private boolean checkIfSameWeightsInMetadata(WeightedRouting newWeights, WeightedRouting oldWeights) {
+        return newWeights.equals(oldWeights);
     }
 
     public void deleteWeightedRoutingMetadata(
         final ClusterDeleteWeightedRoutingRequest request,
         final ActionListener<ClusterDeleteWeightedRoutingResponse> listener
     ) {
+        final long requestVersion = request.getVersion();
+        final String awarenessAttribute = request.getAwarenessAttribute();
         clusterService.submitStateUpdateTask("delete_weighted_routing", new ClusterStateUpdateTask(Priority.URGENT) {
             @Override
             public ClusterState execute(ClusterState currentState) {
                 logger.info("Deleting weighted routing metadata from the cluster state");
+                Metadata metadata = currentState.metadata();
                 Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata());
-                mdBuilder.removeCustom(WeightedRoutingMetadata.TYPE);
+                WeightedRoutingMetadata weightedRoutingMetadata = metadata.custom(WeightedRoutingMetadata.TYPE);
+                ensureNoVersionConflict(requestVersion, weightedRoutingMetadata);
+
+                if ((weightedRoutingMetadata != null && awarenessAttribute == null)
+                    || (weightedRoutingMetadata != null
+                        && weightedRoutingMetadata.getWeightedRouting().attributeName().equals(awarenessAttribute))) {
+                    weightedRoutingMetadata = new WeightedRoutingMetadata(new WeightedRouting(), weightedRoutingMetadata.getVersion() + 1);
```

Comment:
We need to ensure we GC delete this this after sometime.

Reply:
do you mean rather than clearing the weights from metadata instantly on delete request, delete it after x seconds where x can be a dynamic setting?

Reply:
no clear the version in the metadata and reset to -1 after x seconds

Reply:
still a little unclear, is it clear weights from metadata on delete request and update the version, then after x seconds reset the version to -1. What about updates that are made between delete request and version reset?

eg, at t1, delete request updated the version to v+1 and cleared metadata. t1+10 is version reset time
At t1+3 - there is a put call to update weights, so version is now at v+2. 
At t1+10, version is reset to -1 ?

Reply:
Actually you are right, the fact that we could continue with the version without any resource ahead should help us avoid the need to clean up metadata and reset the version, which would otherwise have been needed if it was using up an id

Reply:
Resetting version can have an edge case where an entity can update the weights reading stale weight
eg. someone read the weight with version 1 thinking to update. Meanwhile another entity deleted and updated the weight again which will also have version 1. Now the 1st entity will be able to update the weight, since the service will see version is 1 only